### PR TITLE
Issue #873: Provide a more detailed error if the introspection query's connection fails

### PR DESCRIPTION
**Pull request by marcellomaugeri**
Solve Issue #873

### CODE REVIEW

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/IntrospectiveQuery.kt [hunk 1]
```diff
@@ -61,35 +63,38 @@ class IntrospectiveQuery {
...
                     }
                 """.trimIndent(), MediaType.APPLICATION_JSON_TYPE)
-
-        //TODO check if TCP problems
-        val response = SearchTimeController.measureTimeMillis({ ms, res ->
+        val response: Response?
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/graphql/IntrospectiveQuery.kt:66]: thx. in contrast to Java, in Kotlin most statements are expressions. so can write something directly like `val response = try {...`

Comment (arcuri82) [reply core/src/main/kotlin/org/evomaster/core/problem/graphql/IntrospectiveQuery.kt:66]: also why nullable with `?` ? 
if want to check case for null, could make it non-null, with explicit throw exception after `try` with the `?:` Elvis operator

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/IntrospectiveQuery.kt [hunk 2]
```diff
@@ -61,35 +63,38 @@ class IntrospectiveQuery {
...
-                for (h in list) {
-                    request = request.header(h.first, h.second)
+                    for (h in list) {
+                        request = request.header(h.first, h.second)
+                    }
+                    request.buildPost(query)
+                            .invoke()
+                } catch (e: Exception) {
+                    log.error("Failed query to '$graphQlEndpoint' :  $query")
+                    throw e
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/graphql/IntrospectiveQuery.kt:82]: the `throw SutProblemException(` should rather be done here, with `e.message` added to the message of the error, something like:
```
throw SutProblemException("Failed introspection query to '$graphQlEndpoint', please check connection and URL format. Error: ${e.message}" )
```

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/IntrospectiveQuery.kt [hunk 3]
```diff
@@ -61,35 +63,38 @@ class IntrospectiveQuery {
...
-                    .invoke()
-            } catch (e: Exception) {
-                log.error("Failed query to '$graphQlEndpoint' :  $query")
-                throw e
-            }
-        })
+            })
+        } catch (e: Exception) {
+            throw SutProblemException("Failed introspection query to '$graphQlEndpoint', please check connection and URL format")
+        }
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/graphql/IntrospectiveQuery.kt:87]: not needed here of this outer try/catch. see previous comment

File: core/src/main/kotlin/org/evomaster/core/problem/graphql/IntrospectiveQuery.kt [hunk 4]
```diff
@@ -61,35 +63,38 @@ class IntrospectiveQuery {
...
 
 
 
 
         /*
            Extract the body from response as a string
         */
-        val body = response.readEntity(String::class.java)
+        val body = response?.readEntity(String::class.java)
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/problem/graphql/IntrospectiveQuery.kt:95]: not needed `?` if variable is made non-nullable, with explicit `?:` check on nullability (if needed)

File: core/src/main/kotlin/org/evomaster/core/search/service/SearchTimeController.kt [hunk 5]
```diff
@@ -28,10 +28,13 @@ class SearchTimeController {
...
-            loggingFunction.invoke(System.currentTimeMillis() - startTime, result)
-
-            return result
+            try {
+                val result: T = function.invoke()
+                loggingFunction.invoke(System.currentTimeMillis() - startTime, result)
+                return result
+            } catch(e: Exception) {
+                throw e
+            }
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/service/SearchTimeController.kt:37]: what is the reason of this? why catching an exception and throwing it again without doing anything with it?

**Review by arcuri82 — CHANGES_REQUESTED** [review]
@marcellomaugeri  thx for providing a PR! see my comments

**Comment by marcellomaugeri** [comment]
@arcuri82 Sorry for the delay in troubleshooting. I think it's all sorted now, I was a bit confused as I'm fairly new to Kotlin, thanks for the explanations. However I'm still not sure about the double try-catch to get the answer, could you please check again? Thank you.

**Comment by arcuri82** [comment]
hi @marcellomaugeri 
 sorry, but I don't understand why you are having a double-catch.
Can you please check the branch `related-to-pr-894`, to see if that solves the issue you have? 
thx

**Comment by marcellomaugeri** [comment]
@arcuri82 yes, your branch solves the issue, thank you for your time and patience. I was overcomplicating things! Thank you for your suggestions, I am quite new to Kotlin and I got confused because the code was not compiling and then I added the double-catch, the rest you know

**Pull request closed without merging**